crates/compiler/src/lib.rs

```diff
@@ -186,11 +204,58 @@ macro_rules! pass_instruction {
     };
 }
 
+/// Assigns a value to a variable.
+/// The variable can be stored on the stack or in a register.
+/// If the variable is stored on the stack, the stack pointer will be adjusted in order to store
+/// the value.
+macro_rules! assign_variable {
+    ($codegen:ident, $stack:ident, $location:expr, $value:expr) => {
+        match $location {
+            Location::Stack(offset) => {
+                let offset = -(*offset);
+
+                // adjust the stack pointer
+                $codegen.add_instruction(Instruction::from(Arithmetic::Subtract {
+                    register: Register::Sp,
+                    a: Register::Sp.into(),
+                    b: Number::Int(offset).into(),
+                }));
+
+                // store the result of the expression in the local variable
+                stack_push!($codegen, Register::R0);
+
+                // restore the stack pointer
+                $codegen.add_instruction(Instruction::from(Arithmetic::Add {
+                    register: Register::Sp,
+                    a: Register::Sp.into(),
+                    b: Number::Int(offset).into(),
```

Comment:
review: I don't remember why this doesn't have 1 added to the offset. My belief would be that the `stack_push!` call would increment the stack and cause this to be incorrect. It seems previously this was adjusted by one, but got removed https://github.com/justinrubek/ayysee/blame/29818b224872a99099709e49ec84210ea49a5a0d/crates/compiler/src/lib.rs#L405

Previously:
```rust
                codegen.add_instruction(Instruction::from(Arithmetic::Add {
                    register: Register::Sp.into(),
                    a: Register::Sp.into(),
                    b: Number::Int(*offset as i32 - 1).into(),
                }));
 ```